src/PackageManagement/IDE/DeleteOnRestartManager.cs

```diff
@@ -0,0 +1,187 @@
+        public void MarkPackageDirectoryForDeletion(PackageIdentity package, string packageRoot, INuGetProjectContext projectContext)
+        {
+            if (PackagesFolderPath == null)
+            {
+                return;
+            }
+
+            try
+            {
+                using (FileSystemUtility.CreateFile(packageRoot + DeletionMarkerSuffix, projectContext))
+                {
+                }
+            }
+            catch (Exception e)
+            {
+                projectContext.Log(
+                    MessageLevel.Warning,
+                    string.Format(CultureInfo.CurrentCulture, Strings.Warning_FailedToMarkPackageDirectoryForDeletion, packageRoot, e.Message));
+            }
+        }
+
+        /// <summary>
+        /// Attempts to remove package directories that were unable to be fully deleted during the original uninstall.
+        /// These directories will be marked by an adjacent *directory name*.deleteme files in the local package repository.
+        /// If the directory removal is successful, the .deleteme file will also be removed.
+        /// </summary>
+        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We want to log an exception as a warning and move on")]
+        public void DeleteMarkedPackageDirectories(INuGetProjectContext projectContext)
+        {
+            if (PackagesFolderPath == null)
+            {
+                return;
+            }
+
+            try
+            {
+                var packages = GetPackageDirectoriesMarkedForDeletion();
+                foreach (var package in packages)
+                {
+                    try
+                    {
+                        FileSystemUtility.DeleteDirectorySafe(package, true, projectContext);
+                    }
+                    finally
+                    {
+                        if (!Directory.Exists(package))
+                        {
+                            var deleteMeFilePath = package.TrimEnd('\\') + DeletionMarkerSuffix;
+                            FileSystemUtility.DeleteFile(deleteMeFilePath, projectContext);
+                        }
+                        else
+                        {
+                            projectContext.Log(
+                                MessageLevel.Warning,
+                                string.Format(CultureInfo.CurrentCulture, Strings.Warning_FailedToDeleteMarkedPackageDirectory, package));
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                projectContext.Log(
+                               MessageLevel.Warning,
+                               string.Format(CultureInfo.CurrentCulture, Strings.Warning_FailedToDeleteMarkedPackageDirectory, e.Message));
```

Comment:
This should be using the "Warning_FailedToDeleteMarkedPackageDirector*ies*" resource.